notebooks/02_model/cornac_bpr_deep_dive.ipynb

```diff
@@ -0,0 +1,601 @@
+    "\n",
+    "The implementation of the model is from [Cornac](https://github.com/PreferredAI/cornac), which is a framework for recommender systems with a focus on models leveraging auxiliary data (e.g., item descriptive text and image, social network, etc)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 0 Global Settings and Imports"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "System version: 3.6.8 |Anaconda, Inc.| (default, Feb 21 2019, 18:30:04) [MSC v.1916 64 bit (AMD64)]\n",
+      "Cornac version: 1.1.2\n"
+     ]
+    }
+   ],
+   "source": [
+    "import sys\n",
+    "sys.path.append(\"../../\")\n",
+    "import time\n",
+    "import os\n",
+    "import cornac\n",
+    "import papermill as pm\n",
+    "import pandas as pd\n",
+    "from reco_utils.dataset import movielens\n",
+    "from reco_utils.dataset.python_splitters import python_random_split\n",
+    "from reco_utils.evaluation.python_evaluation import (map_at_k, ndcg_at_k, precision_at_k, recall_at_k)\n",
+    "from reco_utils.recommender.cornac.cornac_utils import predict_ranking\n",
+    "from reco_utils.common.constants import SEED\n",
+    "\n",
+    "print(\"System version: {}\".format(sys.version))\n",
+    "print(\"Cornac version: {}\".format(cornac.__version__))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "tags": [
+     "parameters"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Select MovieLens data size: 100k, 1m, 10m, or 20m\n",
+    "MOVIELENS_DATA_SIZE = '100k'\n",
+    "\n",
+    "# top k items to recommend\n",
+    "TOP_K = 10\n",
+    "\n",
+    "# Model parameters\n",
+    "NUM_FACTORS = 200\n",
+    "NUM_EPOCHS = 100"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1 BPR Algorithm\n",
+    "\n",
+    "### 1.1 Personalized Ranking from Implicit Feedback\n",
+    "\n",
+    "The task of personalized ranking aims at providing each user a ranked list of items (recommendations).  It is very common the scenarios where recommender systems dealing with user implicit behavior (e.g. purchases, clicks).  The available observations are only positive feedback where the non-observed ones are a mixture of real negative feedback and missing values.\n",
+    "\n",
+    "One usual approach for item recommendation is directly predicting a preference score $\\hat{x}_{u,i}$ given to item $i$ by user $u$.  BPR uses a different approach by using item pairs $(i, j)$ and optimizing for the correct ranking given preference of user $u$, thus, there are notions of *positive* and *negative* items.  The training data $D_S : U \\times I \\times I$ is defined as:\n",
+    "\n",
+    "$$D_S := \\{(u, i, j) \\mid i \\in I^{+}_{u} \\wedge j \\in I \\setminus I^{+}_{u}\\}$$\n",
```

Comment:
would you mind to check if the math is rendered correctly? I can't see it on github

Reply:
This is rendered properly on my browser.

Reply:
All the formulas are rendered correctly on GitHub.